test/indexer/test_indexer_controller.py

```diff
@@ -0,0 +1,152 @@
+        sqs = boto3.resource('sqs')
+        for queue_name in config.all_queue_names:
+            sqs.create_queue(QueueName=queue_name)
+
+    def generate_sqs_record(self, **kwargs):
+        event_dict = {
+            'body': json.dumps(kwargs),
+            'receiptHandle': 'ThisWasARandomString',
+            'attributes': {'ApproximateReceiveCount': 1}
+        }
+        context = 'controller_test'
+        return SQSRecord(event_dict=event_dict, context=context)
+
+    @mock_sts
+    def test_contribution_action_invalid(self):
+        event = [self.generate_sqs_record(action='banana', notification='slug', catalog=self.catalog)]
+        self.assertRaises(AssertionError, self.controller.contribute, event)
+
+    @mock_sts
+    def test_contribution_action_reindex(self):
+        event = [self.generate_sqs_record(action='reindex', prefix='ff')]
+        with mock.patch('azul.azulclient.AzulClient.do_remote_reindex') as mock_reindex:
+            mock_reindex.return_value = True
+            self.controller.contribute(event)
+            self.assertTrue(mock_reindex.called)
+            self.assertEqual(mock_reindex.call_count, 1)
+
+    @mock_sts
+    @mock_sqs
+    def test_contribute_and_aggregate(self):
+        """
+        Index two bundles that make contributions to the same project.
+        Inspect that the contributions match the tallies that are returned to
+        SQS. During aggregation only the project entity is deferred due to
+        multiple contributions.
+        """
+        self._create_mock_queues()
+        events = []
+        bundles = []
+        contributions_per_bundle: Sequence[Sequence[Contribution]] = []
+        bundle_fqids = [
+            BundleFQID('56a338fe-7554-4b5d-96a2-7df127a7640b', '2018-03-28T151023.074974Z'),
+            BundleFQID('b2216048-7eaa-45f4-8077-5a3fb4204953', '2018-03-29T104041.822717Z')
+        ]
+        for bundle_fqid in bundle_fqids:
+            notification = self.client.synthesize_notification(self.catalog, bundle_fqid)
+            events.append(self.generate_sqs_record(action='add',
+                                                   catalog=self.catalog,
+                                                   notification=notification))
+            bundle = self._load_canned_bundle(bundle_fqid)
+            bundles.append(bundle)
+            contributions_per_bundle.append(self.index_service.transform(self.catalog, bundle=bundle, delete=False))
+        with mock.patch('azul.indexer.index_controller.IndexController.repository_plugin') as mock_repository, (
+            mock.patch('azul.indexer.index_controller.IndexController.index_service',
+                       new_callable=mock.PropertyMock)) as mock_index_service:
+            mock_index_service.return_value = ForcedRefreshIndexService()
+            mock_repository.return_value = mock.MagicMock()
+            mock_repository.return_value.fetch_bundle.return_value = yield contributions_per_bundle
```

Comment:
everytime that a fetch_bundle is called, we would like to yield a different bundle that was loaded from the canned bundles, this allows us to not actually use the data-plugins to fetch data.

I think I can change this to a `fetch_bundle.side_effect`, it should behave in a similar manor to what we need.
https://docs.python.org/3/library/unittest.mock.html#unittest.mock.Mock.side_effect

I'll try to change this.

Reply:
It was a no-go, by changing this to 
` mock_repository.return_value.fetch_bundle.side_effect = contributions_per_bundle`
The entire list of bundles are returned to where the method call to `fetch_bundle` was made in the `transform`.

I think that we could get around this by doing a more intensive mockup of the `repository_plugin`, but I would like to be precise with the mocks/patches that are made.

Reply:
alright after too much time staring at this problem, i found the issue.
```
mock_repository.return_value = mock.MagicMock()
mock_repository.return_value.fetch_bundle.side_effect = bundles
```
resolves the issue, it was getting a `list` because `contributions_per_bundle` is a list of lists and not what we want to return back to the transform method. >.> 

For some reason when i placed a yield as the return value, the test never really ran, is that what you meant by?

> putting yield here would turn the entire test into a generator and that would definitely not be the behaviour we want.

This was only caught when trying to break the test into sub-tests to show the distinction between the parts of the test, the test output did not show the msg from the sub-tests, upon investigation the tests were passing without running.

Reply:
> For some reason when i placed a yield as the return value, the test never really ran, is that what you meant

Yes, that's what I mean. This SO post give a pretty good explanation of what was going on: https://stackoverflow.com/questions/32128412/python-yield-and-yield-assignment